### ISSUE TITLE
Slow startup when using coc + airline

### ISSUE THREAD

**Issue by devjgm** [neoclide/coc.nvim#1827]
**UPDATE: I updated to the latest neovim, v0.4.3, and the problem persists. New info in the [comment below](https://github.com/neoclide/coc.nvim/issues/1827#issuecomment-620707509).**

## Result from CocInfo

```
## versions
                                                                                                            
vim version: NVIM v0.3.4
node version: v10.15.2
coc.nvim version: 0.0.78-bcb2fe977c
term: xterm
platform: linux
```
## Describe the bug

There's a noticeable startup lag when opening a files with neofim + coc + airline. If I comment out either the coc plugin or the airline plugin, nvim starts up quickly with either one. However, when they are both loaded, there's a noticeable lag when first rendering the screen.

## Reproduce the bug

**We will close your issue when you don't provide minimal vimrc and we can't
reproduce it**

- Create file `init.vim` with：

```vim
profile start profile.log
profile func *
profile file *
call plug#begin(stdpath('data') . '/plugged')
Plug 'vim-airline/vim-airline'
Plug 'neoclide/coc.nvim', {'branch': 'release'}
call plug#end()
```

- Start (neo)vim with command: `nvim -u init.vim`

Make sure plugins are installed with `:PlugInstall`

- Operate vim.

There's nothing to do here. The performance lag is just opening nvim and waiting for the screen to draw.

## Output from my profile.log

I don't know what `<SNR>16_request()` does, but it looks like maybe that's suggestive of the problem.

```
...
FUNCTION  <SNR>16_request()
Called 3 times
Total time:   0.605219
 Self time:   0.605019

count  total (s)   self (s)
    3   0.000116   0.000068   let channel = coc#client#get_channel(self)
    3              0.000030   if empty(channel) | return '' | endif
    3              0.000012   try
    3              0.000010     if s:is_vim
                                  let res = ch_evalexpr(channel, [a:method, a:args], {'timeout': 60 * 1000})
                                  if type(res) == 1 && res ==# ''
                                    throw 'request '.a:method. ' '.string(a:args).' timeout after 60s'
                                  endif
                                  let [l:errmsg, res] =  res
                                  if !empty(l:errmsg)
                                    throw l:errmsg
                                  else
                                    return res
                                  endif
                                else
    3   0.604691   0.604538       return call('rpcrequest', [channel, a:method] + a:args)
                                endif
                              catch /.*/
                                if v:exception =~# 'E475'
                                  if get(g:, 'coc_vim_leaving', 0) | return | endif
                                  echohl Error | echom '['.self.name.'] server connection lost' | echohl None
                                  let name = self.name
                                  call s:on_exit(name, 0)
                                  execute 'silent do User ConnectionLost'.toupper(name[0]).name[1:]
                                elseif v:exception =~# 'E12'
                                  " neovim's bug, ignore it
                                else
                                  echohl Error | echo 'Error on request ('.a:method.'): '.v:exception | echohl None
                                endif
                              endtry
...

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    3   0.605219   0.605019  <SNR>16_request()
  218   0.021931   0.010474  airline#highlighter#get_highlight()
    2   0.032725   0.010130  <SNR>9_init()
  436              0.009780  <SNR>36_get_syn()
  138   0.030876   0.007830  airline#highlighter#exec()
    4   0.037661   0.006919  airline#highlighter#highlight()
   71              0.005316  <SNR>36_GetHiCmd()
    1   0.006847   0.004474  airline#extensions#load()
   24   0.027309   0.003199  airline#check_mode()
   24   0.003139   0.002538  airline#extensions#whitespace#check()
    1              0.002411  <SNR>13_AddAnsiGroups()
   10   0.003127   0.002362  <SNR>35_create()
  138              0.001936  <SNR>36_CheckDefined()
  218              0.001677  <SNR>36_get_array()
    1              0.001674  <SNR>16_start()
    2   0.012383   0.001575  12()
  184              0.001422  airline#util#winwidth()
   24   0.562112   0.001324  airline#extensions#hunks#get_hunks()
  168              0.001315  airline#util#append()
    3   0.002940   0.001192  airline#init#bootstrap()

```

**Comment by fannheyward** [comment]
Upgrade your neovim to latest version and try again.

**Comment by devjgm** [comment]
OK. I updated to the latest, neovim v0.4.3. The problem persists. Here's the latest info:

## CocInfo

```
## versions

vim version: NVIM v0.4.3
node version: v10.15.2
coc.nvim version: 0.0.78-bcb2fe977c
term: xterm
platform: linux
```

## Snippets from `profile.log`

```
...
FUNCTION  <SNR>16_request()
    Defined: ~/.local/share/nvim/plugged/coc.nvim/autoload/coc/client.vim line 122
Called 3 times
Total time:   0.535039
 Self time:   0.534971

count  total (s)   self (s)
    3   0.000083   0.000049   let channel = coc#client#get_channel(self)
    3              0.000019   if empty(channel) | return '' | endif
    3              0.000006   try
    3              0.000006     if s:is_vim
                                  let res = ch_evalexpr(channel, [a:method, a:args], {'timeout': 60 * 1000})
                                  if type(res) == 1 && res ==# ''
                                    throw 'request '.a:method. ' '.string(a:args).' timeout after 60s'
                                  endif
                                  let [l:errmsg, res] =  res
                                  if !empty(l:errmsg)
                                    throw l:errmsg
                                  else
                                    return res
                                  endif
    3              0.000004     else
    3   0.534645   0.534611       return call('rpcrequest', [channel, a:method] + a:args)
                                endif
                              catch /.*/
                                if v:exception =~# 'E475'
                                  if get(g:, 'coc_vim_leaving', 0) | return | endif
                                  echohl Error | echom '['.self.name.'] server connection lost' | echohl None
                                  let name = self.name
                                  call s:on_exit(name, 0)
                                  execute 'silent do User ConnectionLost'.toupper(name[0]).name[1:]
                                elseif v:exception =~# 'E12'
                                  " neovim's bug, ignore it
                                else
                                  echohl Error | echo 'Error on request ('.a:method.'): '.v:exception | echohl None
                                endif
    3              0.000007   endtry

...
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    3   0.535039   0.534971  <SNR>16_request()
  218   0.032065   0.015951  airline#highlighter#get_highlight()
  436              0.013826  <SNR>37_get_syn()
  138   0.046233   0.012812  airline#highlighter#exec()
    2   0.036125   0.011314  <SNR>9_init()
    4   0.061315   0.010266  airline#highlighter#highlight()
   71              0.007114  <SNR>37_GetHiCmd()
    1   0.007380   0.004790  airline#extensions#load()
   23   0.048596   0.002960  airline#check_mode()
  138              0.002913  <SNR>37_CheckDefined()
   10   0.003696   0.002818  <SNR>36_create()
   23   0.003254   0.002583  airline#extensions#whitespace#check()
  218              0.002288  <SNR>37_get_array()
    1              0.002147  <SNR>13_AddAnsiGroups()
  161              0.001522  airline#util#append()
   23   0.509760   0.001514  airline#extensions#hunks#get_hunks()
   28   0.017747   0.001514  <SNR>37_exec_separator()
  177              0.001434  airline#util#winwidth()
   69   0.001865   0.001260  airline#util#shorten()
    2   0.010302   0.001243  12()
```

**Comment by devjgm** [comment]
Also, here's the output of `nvim --startuptime`. The bottom of the file shows 590ms on "first screen update".

```
times in msec
 clock   self+sourced   self:  sourced script
 clock   elapsed:              other lines

000.031  000.031: --- NVIM STARTING ---
000.684  000.654: locale set
001.792  001.108: inits 1
001.834  000.041: window checked
002.255  000.422: parsing arguments
002.694  000.438: expanding arguments
002.847  000.153: inits 2
004.869  002.023: init highlight
010.880  005.373  005.373: sourcing /home/jgm/.local/share/nvim/site/autoload/plug.vim
029.568  015.887  015.887: sourcing /usr/share/nvim/runtime/filetype.vim
029.849  000.073  000.073: sourcing /usr/share/nvim/runtime/ftplugin.vim
030.040  000.060  000.060: sourcing /usr/share/nvim/runtime/indent.vim
030.794  000.376  000.376: sourcing /usr/share/nvim/runtime/syntax/syncolor.vim
030.963  000.682  000.306: sourcing /usr/share/nvim/runtime/syntax/synload.vim
031.011  000.864  000.182: sourcing /usr/share/nvim/runtime/syntax/syntax.vim
031.037  025.766  003.510: sourcing /home/jgm/.config/nvim/init.vim
031.043  000.407: sourcing vimrc file(s)
032.038  000.173  000.173: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/init.vim
032.734  000.174  000.174: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/parts.vim
034.869  000.340  000.340: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/util.vim
035.008  003.576  002.888: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/plugin/airline.vim
035.541  000.212  000.212: sourcing /home/jgm/.local/share/nvim/plugged/coc.nvim/autoload/coc/rpc.vim
036.399  000.740  000.740: sourcing /home/jgm/.local/share/nvim/plugged/coc.nvim/autoload/coc/util.vim
036.998  000.272  000.272: sourcing /home/jgm/.local/share/nvim/plugged/coc.nvim/autoload/coc/client.vim
041.451  006.260  005.036: sourcing /home/jgm/.local/share/nvim/plugged/coc.nvim/plugin/coc.vim
042.444  000.442  000.442: sourcing /usr/share/nvim/runtime/plugin/gzip.vim
042.525  000.019  000.019: sourcing /usr/share/nvim/runtime/plugin/health.vim
042.807  000.244  000.244: sourcing /usr/share/nvim/runtime/plugin/man.vim
043.373  000.523  000.523: sourcing /usr/share/nvim/runtime/plugin/matchit.vim
043.645  000.200  000.200: sourcing /usr/share/nvim/runtime/plugin/matchparen.vim
044.505  000.808  000.808: sourcing /usr/share/nvim/runtime/plugin/netrwPlugin.vim
044.827  000.212  000.212: sourcing /usr/share/nvim/runtime/plugin/rplugin.vim
045.105  000.227  000.227: sourcing /usr/share/nvim/runtime/plugin/shada.vim
045.181  000.034  000.034: sourcing /usr/share/nvim/runtime/plugin/spellfile.vim
045.461  000.239  000.239: sourcing /usr/share/nvim/runtime/plugin/tarPlugin.vim
045.589  000.085  000.085: sourcing /usr/share/nvim/runtime/plugin/tohtml.vim
045.653  000.026  000.026: sourcing /usr/share/nvim/runtime/plugin/tutor.vim
045.890  000.193  000.193: sourcing /usr/share/nvim/runtime/plugin/zipPlugin.vim
045.898  001.769: loading plugins
046.001  000.103: loading packages
046.050  000.050: loading after plugins
046.065  000.015: inits 3
047.485  001.420: reading ShaDa
048.885  001.400: clearing screen
049.879  000.343  000.343: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions.vim
050.131  000.078  000.078: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/quickfix.vim
050.452  000.217  000.217: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline.vim
050.643  000.042  000.042: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/netrw.vim
050.978  000.076  000.076: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/section.vim
051.332  000.194  000.194: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/highlighter.vim
051.509  000.752  000.482: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/term.vim
051.764  000.106  000.106: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/hunks.vim
052.072  000.055  000.055: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/coc.vim
052.380  000.127  000.127: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/whitespace.vim
052.904  000.095  000.095: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/wordcount.vim
053.090  000.031  000.031: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/keymap.vim
061.148  000.064  000.064: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/themes.vim
061.467  000.504  000.440: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/themes/dark.vim
078.916  000.137  000.137: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/builder.vim
079.265  000.073  000.073: sourcing /home/jgm/.local/share/nvim/plugged/vim-airline/autoload/airline/extensions/default.vim
090.821  039.376: opening buffers
090.881  000.060: BufEnter autocommands
090.886  000.005: editing files in windows
091.389  000.504: VimEnter autocommands
091.404  000.014: before starting main loop
681.772  590.368: first screen update
681.779  000.007: --- NVIM STARTED ---
```

**Comment by chemzqm** [comment]
We need debug log of coc.nvim to address this issue, please follow these steps:

* Add `let g:node_client_debug = 1` to your vimrc.
* Restart your vim/neovim to make the issue happen.
* Open the log file by command `:call coc#client#open_log()`.
* Attach the log file or paste content from the log.

It should not use request at all.

**Comment by chemzqm** [comment]
It's problem of `CocAction` at https://github.com/vim-airline/vim-airline/blob/master/autoload/airline/extensions/hunks.vim#L25, it could be slow on startup

**Comment by oblitum** [comment]
I recommend lightline (or something lighter or your own statusline). I've hit many of such issues with airline in the past.

**Comment by devjgm** [comment]
@chemzqm thank you. I think you identified the problem correctly. Indeed, if I `let g:airline#extensions#hunks#enabled = 0` then the problem goes away. I'll do this for now.

@oblitum thanks for the lightline suggestion. I do like lightline. But the thing that keeps me with airline is the nice looking tabline across the top that shows _buffers_ rather than tabs. I know there are ways to make that work with lightline, and maybe I'll do that if airline continues to cause problems.

**Comment by chrisbra** [comment]
> 
> 
> It's problem of `CocAction` at https://github.com/vim-airline/vim-airline/blob/master/autoload/airline/extensions/hunks.vim#L25, it could be slow on startup.
> Remove that code or request for an option to disable detect.

that is funny that you mention this here but not to me, when I asked about it https://github.com/neoclide/coc-git/issues/96

Also you did not come back to me with the issues I had about it, simply closed it...

**Comment by chemzqm** [comment]
@chrisbra I've answered your question before close the issue. One possible solution is use `CocActionAsync` or check if the extension exists before `VimEnter`, it's expected to be slow since coc.nvim load all extensions in asynchronous manner.

**Comment by chrisbra** [comment]
You did not come back to why it did not work. Anyhhow, what exactly do you mean by:

> CocActionAsync or check if the extension exists before VimEnter

Or asked the other way around, what would be the best way to ask from another vimplugin to check for coc-git extension being loaded.

**Comment by chemzqm** [comment]
> what would be the best way to ask from another vimplugin to check for coc-git extension being loaded.

My preferred way is not ask at all, provide an option then use `g:coc_git_status` if user change the option to true, the variable could be never exists when user change `git.enableGlobalStatus` configuration to `false`

**Comment by oblitum** [comment]
> that is funny that you mention this here but not to me, when I asked about it [neoclide/coc-git#96](https://github.com/neoclide/coc-git/issues/96)

The knowledge of `CocAction` vs `CocActionAsync` is pretty widespread here, to the point when someone refers to one, it's simply assumed that a whole exposition of the two isn't necessary. If you need things async, then just use async version.

Aside that, "air"line started being as a light alternative to powerline but simply lost its way with myriad integrations which are not fully tested. Also as usual, same bad tone when dealing with problems.

**Comment by chrisbra** [comment]
> The knowledge of CocAction vs CocActionAsync is pretty widespread here, to the point when someone refers to one, it's simply assumed that a whole exposition of the two isn't necessary. If you need things async, then just use async version.

While that might be true, it is pretty obvious I am not using coc. And trying to dive into the asycs of another vim plugin might be a bit much asked for.

> Aside that, "air"line started being as a light alternative to powerline but simply lost its way with myriad integrations which are not fully tested. 

If you have improvements for better integrations, I am all ears. I have never anything against concrete suggestions.

> Also as usual, same bad tone when dealing with problems.

You mean from your side, as usual? Talk for yourself please. I am trying to be helpful. Thanks for understanding.

**Comment by chrisbra** [comment]
> My preferred way is not ask at all, provide an option then use g:coc_git_status if user change the option to true, the variable could never exists when user change git.enableGlobalStatus configuration to false

My preferred way would be to use whatever is available, without having the user asked and configured something, that can be configured automatically. After all, that's what computer are good for.

**Comment by devjgm** [comment]
![homer](https://user-images.githubusercontent.com/9447643/80607944-18df2180-8a04-11ea-9da8-09734dd9df07.gif)

**Comment by oblitum** [comment]
> You mean from your side, as usual? Talk for yourself please. I am trying to be helpful. Thanks for understanding.

No, I mean "**that is funny** that you mention this here but not to me, when I asked about it neoclide/coc-git#96". It isn't polite. And you started with sarcasm.

My last message on this off-topic.

**Comment by chrisbra** [comment]
Look, I asked explicitly how to check for the availability of the coc-git extension from another plugin, because that's what has been asked for in vim-airline. There was no mentioning that this could be slow.  How should I know that this is known to be slow when providing this feature? Then mentioning it here in just another issue, where I do not notice it is not helpful. Just ping me if you know this is to be slow or notice it here. I am pretty reactive and even notice if you ping me in other issues.

> And you started with sarcasm.

Yes, I asked politely for how to check it. At the very least, one could mention it. Is this asked too much?  Yes, I am frustrated, when I notice that issues of vim-airline are being discussed elsewhere where I do not have to a chance to notice and fix it. And yes, this includes your message:

>  I've hit many of such issues with airline in the past, unrelated to coc.nvim. 

as well. If you notice a problem, mention it. I'll try to fix it if it is feasible and I have time and that's how I handle all of my plugin or wherever I take part in a community.

**Comment by chrisbra** [comment]
@devjgm  should be fixed as of https://github.com/vim-airline/vim-airline/commit/48064e24d5d1fd715b2366444a09475bcbe03879

**Comment by chemzqm** [comment]
> My preferred way would be to use whatever is available, without having the user asked and configured something, that can be configured automatically. After all, that's what computer are good for.

The detection can be slow, so an option can be provided to avoid it  like what neovim does for `g:python_host_prog`

**Comment by chrisbra** [comment]
please correct me if I am wrong, but for once I believe there were also Neovim issues for it being slow and second, since this is async, it may not be available right from the time vim is started and the statusline is drawn the first time. 

Anyhow, I have now removed the whole check in airline.

**Comment by chemzqm** [comment]
CocActionAsync can be used for none block function call to coc.nvim, but that may require lots of change to your codebase.
